internal/headerbp/headerbp.go

```diff
@@ -0,0 +1,355 @@
+package headerbp
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+)
+
+const (
+	// headerPrefixCanonicalHTTP and headerPrefixLower must remain the same character length
+	headerPrefixCanonicalHTTP = "X-Bp-"
+	headerPrefixLower         = "x-bp-"
+
+	IsUntrustedRequestHeaderCanonicalHTTP = "X-Rddt-Untrusted"
+	IsUntrustedRequestHeaderLower         = "x-rddt-untrusted"
+)
+
+var SetV2Context = func(ctx context.Context, _ map[string]string) context.Context {
+	return ctx
+}
+
+// ErrNewInternalHeaderNotAllowed is returned by a client when the call tries to set an internal header is not allowlisted
+var ErrNewInternalHeaderNotAllowed = fmt.Errorf("cannot send new internal headers on requests")
+
+// IsBaseplateHeader returns true if the header is for baseplate and should be propagated
+func IsBaseplateHeader(key string) bool {
+	if len(key) < len(headerPrefixLower) {
+		return false
+	}
+	prefix := key[:len(headerPrefixLower)]
+	return prefix == headerPrefixLower || prefix == headerPrefixCanonicalHTTP
```

Comment:
nit: just use:
```suggestion
	return strings.HasPrefix(key, headerPrefixLower) || strings.HasPrefix(key, headerPrefixCanonicalHTTP)
```
?

this way you no longer rely on `headerPrefixCanonicalHTTP` and `headerPrefixLower` being the same length.

Reply:
the `strings.HasPrefix` approach is significantly slower (relatively speaking)

BenchmarkIsBaseplateHeader
BenchmarkIsBaseplateHeader/startswith
BenchmarkIsBaseplateHeader/startswith-16         	299018905	         3.799 ns/op	       0 B/op	       0 allocs/op
BenchmarkIsBaseplateHeader/==
BenchmarkIsBaseplateHeader/==-16                 	1000000000	         0.2518 ns/op	       0 B/op	       0 allocs/op
PASS

Reply:
> this way you no longer rely on headerPrefixCanonicalHTTP and headerPrefixLower being the same length.

They must always be the same since they are just the same string with different cases